AX: Flakey failures in wpt.fyi Interop 2024 accessibility tests

twilco opened the pull request:
#### 4900d7a90e6ea0edfd43a2bfdf37db22ad682170
<pre>
AX: Flakey failures in wpt.fyi Interop 2024 accessibility tests
<a href="https://bugs.webkit.org/show_bug.cgi?id=278540">https://bugs.webkit.org/show_bug.cgi?id=278540</a>
<a href="https://rdar.apple.com/134527513">rdar://134527513</a>

Reviewed by Chris Fleizach.

We suspect these flakey failures happen if WebDriver queries WebCore before the render tree has been built for the first
time, meaning we aren&apos;t going to create accessibility objects correctly, resulting in an empty string result for computed
role and computed label.

We actually worked around this problem in the implementation of test_driver_internal.{get_computed_label, get_computed_role}
by using a mechanism to asynchronously wait for the render tree to be laid out via `Internals::readyToRetrieveComputedRoleOrLabel`.
However, no such mechanism was put in place for the actual WebDriver implementation of getComputedRole and getComputedLabel,
so it is probably suffering from the same problem.

With this patch, `Internals::readyToRetrieveComputedRoleOrLabel` is removed, and instead we just force layout to ensure
renderers are attached to elements before querying these properties. We also now force layout in the WebDriver methods
for this functionality in WebAutomationSessionProxy.cpp, which should solve the flakiness.

* LayoutTests/imported/w3c/web-platform-tests/resources/testdriver-vendor.js:
* Source/WebCore/accessibility/AccessibilityObject.cpp:
(WebCore::AccessibilityObject::computedLabel):
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::getComputedLabel const):
(WebCore::Internals::getComputedRole const):
(WebCore::Internals::readyToRetrieveComputedRoleOrLabel const): Deleted.
* Source/WebCore/testing/Internals.h:
* Source/WebCore/testing/Internals.idl:
* Source/WebKit/WebProcess/Automation/WebAutomationSessionProxy.cpp:
(WebKit::WebAutomationSessionProxy::getAccessibilityObjectForNode):
(WebKit::WebAutomationSessionProxy::getComputedRole):
(WebKit::WebAutomationSessionProxy::getComputedLabel):

Canonical link: <a href="https://commits.webkit.org/282659@main">https://commits.webkit.org/282659@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/92842c0b07d700e9e43f6887b1e0ee2afc11ed5e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/63777 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/43134 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/16374 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/67798 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/14385 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/65897 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/50824 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/14665 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/51394 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/9948 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/66846 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/39959 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/55209 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/32009 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/36639 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/12587 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/13258 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/58476 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/12914 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/69494 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/7724 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/12471 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/58717 "Found 1 new test failure: fast/forms/search/search-size-with-decorations.html (failure)") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/7757 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/55303 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/58864 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/14124 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/6425 "Passed tests") | | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/38954 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/40033 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/41216 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/39776 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

fleizach commented on Source/WebCore/testing/Internals.cpp:
```diff
@@ -7462,27 +7462,22 @@ AccessibilityObject* Internals::axObjectForElement(Element& element) const
-    if (auto* axObject = axObjectForElement(element))
-        return axObject->computedRoleString();
-    return ""_s;
+    RefPtr axObject = axObjectForElement(element);
```
are role and label the only things that are impacted?

Should we move 
if (RefPtr document = contextDocument())
        document->updateLayoutIgnorePendingStylesheets();

into a standard function that can be called by any methods that need to do this

twilco replied:
> Should we move ... into a standard function that can be called by any methods that need to do this

Fixed!

> are role and label the only things that are impacted?

Yes, these are the only two properties we need for the WPTs at this point.

nt1m commented on Source/WebKit/WebProcess/Automation/WebAutomationSessionProxy.cpp:
```diff
@@ -330,7 +330,7 @@ WebCore::AccessibilityObject* WebAutomationSessionProxy::getAccessibilityObjectF
     }
 
-    WebCore::Element* coreElement = elementForNodeHandle(*frame, nodeHandle);
+    RefPtr<WebCore::Element> coreElement = elementForNodeHandle(*frame, nodeHandle);
```
```suggestion
    RefPtr coreElement = elementForNodeHandle(*frame, nodeHandle);
```

nt1m commented on Source/WebKit/WebProcess/Automation/WebAutomationSessionProxy.cpp:
```diff
@@ -339,9 +339,13 @@ WebCore::AccessibilityObject* WebAutomationSessionProxy::getAccessibilityObjectF
-    if (WebCore::AXObjectCache* axObjectCache = coreElement->document().axObjectCache()) {
-        if (WebCore::AccessibilityObject* axObject = axObjectCache->getOrCreate(coreElement))
-            return axObject;
+    if (CheckedPtr<WebCore::AXObjectCache> axObjectCache = coreElement->document().axObjectCache()) {
```
```suggestion
    if (CheckedPtr axObjectCache = coreElement->document().axObjectCache()) {
```

nt1m commented on Source/WebCore/accessibility/AccessibilityObject.cpp:
```diff
@@ -352,7 +352,7 @@ String AccessibilityObject::computedLabel()
     // This method is being called by WebKit inspector, which may happen at any time, so we need to update our backing store now.
     // Also hold onto this object in case updateBackingStore deletes this node.
-    RefPtr<AccessibilityObject> protectedThis(this);
+    Ref protectedThis { *this };
```
(probably a separate change, but protectors are not a recommended pattern anymore, the caller should be protecting)

twilco committed:
AX: Flakey failures in wpt.fyi Interop 2024 accessibility tests
https://bugs.webkit.org/show_bug.cgi?id=278540
rdar://134527513

Reviewed by Chris Fleizach.

We suspect these flakey failures happen if WebDriver queries WebCore before the render tree has been built for the first
time, meaning we aren't going to create accessibility objects correctly, resulting in an empty string result for computed
role and computed label.

We actually worked around this problem in the implementation of test_driver_internal.{get_computed_label, get_computed_role}
by using a mechanism to asynchronously wait for the render tree to be laid out via `Internals::readyToRetrieveComputedRoleOrLabel`.
However, no such mechanism was put in place for the actual WebDriver implementation of getComputedRole and getComputedLabel,
so it is probably suffering from the same problem.

With this patch, `Internals::readyToRetrieveComputedRoleOrLabel` is removed, and instead we just force layout to ensure
renderers are attached to elements before querying these properties. We also now force layout in the WebDriver methods
for this functionality in WebAutomationSessionProxy.cpp, which should solve the flakiness.

* LayoutTests/imported/w3c/web-platform-tests/resources/testdriver-vendor.js:
* Source/WebCore/accessibility/AccessibilityObject.cpp:
(WebCore::AccessibilityObject::computedLabel):
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::getComputedLabel const):
(WebCore::Internals::getComputedRole const):
(WebCore::Internals::readyToRetrieveComputedRoleOrLabel const): Deleted.
* Source/WebCore/testing/Internals.h:
* Source/WebCore/testing/Internals.idl:
* Source/WebKit/WebProcess/Automation/WebAutomationSessionProxy.cpp:
(WebKit::WebAutomationSessionProxy::getAccessibilityObjectForNode):
(WebKit::WebAutomationSessionProxy::getComputedRole):
(WebKit::WebAutomationSessionProxy::getComputedLabel):

Canonical link: https://commits.webkit.org/282659@main
